File: scripts/analysis/unreject_submissions.py
#!/usr/bin/env python3
"""
Unreject previously rejected submissions (REJECTED → AWAITING REVIEW).

Python port of scripts/unreject-submissions.ts.

Used when a rejection should be reversed so the participant can be offered
a voluntary return instead.

Environment variables:
  PROLIFIC_API_TOKEN  – Prolific API token (required)
  STUDY_ID            – Prolific study ID (required)
  PID_LIST            – Comma-separated PIDs to unreject (required)
  DRY_RUN             – When "false", unreject live (default: true)
"""

import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from tabs_api import prolific_submissions, prolific_unreject


def _require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        print(f"Error: {name} is required", file=sys.stderr)
        sys.exit(1)
    return value


def main():
    api_token = _require_env("PROLIFIC_API_TOKEN")
    study_id = _require_env("STUDY_ID")
    pid_list_raw = _require_env("PID_LIST")
    dry_run = os.environ.get("DRY_RUN", "true").strip().lower() != "false"

    pids = [p.strip() for p in pid_list_raw.split(",") if p.strip()]

    print("================================================================")
    print("  Unreject Submissions (REJECTED → AWAITING REVIEW)")
    print(f"  Study: {study_id}")
    print(f"  PIDs: {len(pids)}")
    print(f"  Mode: {'DRY RUN' if dry_run else 'LIVE'}")
    print("================================================================")
    print()

    # Look up submission IDs
    subs = prolific_submissions(study_id, api_token)
    pid_to_sub: dict[str, dict] = {}
    for s in subs:
        pid_to_sub[s.get("participant_id", "")] = {
            "id": s.get("id", ""),
            "status": s.get("status", ""),
        }

    unrejected = 0
    would_unreject = 0
    skipped = 0
    not_found = 0
    failed = 0

    for pid in pids:
        sub = pid_to_sub.get(pid)
        if not sub:
            print(f"  NOT FOUND: {pid}")
            not_found += 1
            continue
        if sub["status"] != "REJECTED":
            print(f"  SKIP: {pid} - status is {sub['status']} (not REJECTED)")
            skipped += 1
            continue

        if dry_run:
            print(f"  DRY RUN: would unreject {pid} (submission {sub['id']})")
            would_unreject += 1
        else:
            try:
                prolific_unreject(sub["id"], api_token)
                unrejected += 1
            except Exception as e:
                print(f"  FAILED: {pid} - {e}")
                failed += 1

    print()
    if dry_run:
        print(
            f"WOULD UNREJECT: {would_unreject} | SKIPPED: {skipped} "
            f"| NOT FOUND: {not_found}"
        )
    else:
        print(
            f"UNREJECTED: {unrejected} | SKIPPED: {skipped} "
            f"| NOT FOUND: {not_found} | FAILED: {failed}"
        )


if __name__ == "__main__":
    main()

File: scripts/archive/unreject-submissions.ts
/**
 * Unreject previously rejected submissions (transition REJECTED → AWAITING REVIEW).
 *
 * Used when a rejection should be reversed so the participant can be offered
 * a voluntary return instead.
 *
 * Environment:
 *   PROLIFIC_API_TOKEN  – Prolific API token (required)
 *   STUDY_ID            – Prolific study ID (required)
 *   PID_LIST            – Comma-separated PIDs to unreject (required)
 *   DRY_RUN             – When "false", unreject live (default: true)
 */

import { getCurrentUser, listStudySubmissions, unrejectSubmission } from '../src/lib/prolific-api'

async function main() {
  const token = process.env.PROLIFIC_API_TOKEN
  if (!token) {
    console.error('PROLIFIC_API_TOKEN required')
    process.exit(1)
  }
  const studyId = process.env.STUDY_ID
  if (!studyId) {
    console.error('STUDY_ID required')
    process.exit(1)
  }
  const pidListRaw = process.env.PID_LIST
  if (!pidListRaw) {
    console.error('PID_LIST required')
    process.exit(1)
  }
  const dryRun = (process.env.DRY_RUN ?? 'true').toLowerCase() !== 'false'

  const pids = pidListRaw
    .split(',')
    .map((p) => p.trim())
    .filter(Boolean)

  console.log('================================================================')
  console.log('  Unreject Submissions (REJECTED → AWAITING REVIEW)')
  console.log(`  Study: ${studyId}`)
  console.log(`  PIDs: ${pids.length}`)
  console.log(`  Mode: ${dryRun ? 'DRY RUN' : 'LIVE'}`)
  console.log('================================================================')
  console.log('')

  const user = await getCurrentUser(token)
  console.log(`Connected as: ${user.name} (${user.email})`)
  console.log('')

  // Look up submission IDs
  const subs = await listStudySubmissions(studyId, token)
  const pidToSub = new Map<string, { id: string; status: string }>()
  for (const s of subs.results || []) {
    pidToSub.set(s.participant_id, { id: s.id, status: s.status })
  }

  let unrejected = 0
  let skipped = 0
  let notFound = 0
  let failed = 0

  for (const pid of pids) {
    const sub = pidToSub.get(pid)
    if (!sub) {
      console.log(`  NOT FOUND: ${pid}`)
      notFound++
      continue
    }
    if (sub.status !== 'REJECTED') {
      console.log(`  SKIP: ${pid} - status is ${sub.status} (not REJECTED)`)
      skipped++
      continue
    }

    if (dryRun) {
      console.log(`  DRY RUN: would unreject ${pid} (submission ${sub.id})`)
      unrejected++
    } else {
      try {
        await unrejectSubmission(sub.id, token)
        unrejected++
      } catch (error: unknown) {
        const msg = error instanceof Error ? error.message : String(error)
        console.error(`  FAILED: ${pid} - ${msg}`)
        failed++
      }
    }
  }

  console.log('')
  console.log(
    `UNREJECTED: ${unrejected} | SKIPPED: ${skipped} | NOT FOUND: ${notFound} | FAILED: ${failed}`
  )
}

main().catch((e) => {
  console.error(e)
  process.exit(1)
})
